Fix and Enable manual tests in Jenkins

mellery451 opened the pull request:
@seelabs I included your change from https://github.com/ripple/rippled/pull/2497 in order to get all the manual tests to pass. If you want to close that PR, I think we can keep the change here as a separate commit (your call). I also changed the way test suite order prioritization is handled and would appreciate your feedback on that. 

If you're wondering how I came up with the priority numbers declared for various tests: I timed each long-running/manual test and then multiplied by 10. For anything under a minute, I just picked some reasonable number under 10. This can certainly be revisited if we need to fine-tune the order of execution.

ripplelabs-jenkins commented:
## Jenkins Build Summary

Built from [this commit](https://github.com/ripple/rippled/commit/6c629e3dd76278fb89e0c0b22356f597763c14e5)

Built at __20180509 - 22:58:10__

### Test Results

Build Type | Log | Result | Status
---------- | --- | ------ | ------
coverage | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/coverage.txt) | 1020 cases, 0 failed, t: 619s | PASS :white_check_mark: 
docs | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/docs.txt) | 1 cases, 0 failed, t: 2s | PASS :white_check_mark: 
gcc.release.unity <br/>-Dassert=true, <br/>MANUAL_TESTS=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/gcc.release.unity_Dassert_true_MANUAL_TESTS_true.txt) | 1093 cases, 0 failed, t: 783s | PASS :white_check_mark: 
clang.debug.unity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/clang.debug.unity.txt) | 1020 cases, 0 failed, t: 249s | PASS :white_check_mark: 
clang.debug.nounity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/clang.debug.nounity.txt) | 1018 cases, 0 failed, t: 146s | PASS :white_check_mark: 
clang.debug.unity, <br/>PARALLEL_TESTS=false | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/clang.debug.unity_PARALLEL_TESTS_false.txt) | 1020 cases, 0 failed, t: 486s | PASS :white_check_mark: 
gcc.debug.unity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/gcc.debug.unity.txt) | 1020 cases, 0 failed, t: 243s | PASS :white_check_mark: 
gcc.debug.nounity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/gcc.debug.nounity.txt) | 1018 cases, 0 failed, t: 101s | PASS :white_check_mark: 
clang.release.unity <br/>-Dassert=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/clang.release.unity_Dassert_true.txt) | 1019 cases, 0 failed, t: 329s | PASS :white_check_mark: 
gcc.release.unity <br/>-Dassert=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/gcc.release.unity_Dassert_true.txt) | 1019 cases, 0 failed, t: 354s | PASS :white_check_mark: 
gcc.debug.unity <br/>-Dstatic=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/gcc.debug.unity_Dstatic_true.txt) | 1020 cases, 0 failed, t: 243s | PASS :white_check_mark: 
msvc.debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/msvc.debug.txt) | 1017 cases, 0 failed, t: 561s | PASS :white_check_mark: 
msvc.debug, <br/>PROJECT_NAME=rippled_classic | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/msvc.debug_PROJECT_NAME_rippled_classic.txt) | 1017 cases, 0 failed, t: 1169s | PASS :white_check_mark: 
msvc.release | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2517/1//artifact/msvc.release.txt) | 1016 cases, 0 failed, t: 504s | PASS :white_check_mark:

mellery451 commented:
@scottschurr -- since you've expressed in interest in manual tests in the past, I welcome your feedback as well.

seelabs commented:
@mellery451 I closed the PR with the manual offer test fix; we'll make it part of this one.

scottschurr commented:
I see a number of the files that are changed are in the `src/beast/extras` area.  I think that's the pre-Boost Beast sub-tree.  I'm guessing that the thought here is that pre-Boost Beast will never see any maintenance, so we're free to modify as we want.  But I'd like to talk about that explicitly rather than make personal assumptions.

Thoughts?

mellery451 commented:
@scottschurr ...these files are not included in the installed boost::beast directories (although they are somewhere in the official repo), and I noticed that @seelabs left this directory in our tree when implementing https://github.com/ripple/rippled/pull/2485, so I felt somewhat assured in changing the files. That said, I agree with your concern and would appreciate some more clarity about how we want to handle these sources after moving to boost::beast.

seelabs commented on src/beast/extras/beast/unit_test/suite_info.hpp:
```diff
@@ -87,9 +90,12 @@ class suite_info
+                   std::tie(rhs.library_, rhs.module_, rhs.name_);
+        else
+            // we want higher priority suites sorted first
+            return lhs.priority_ > rhs.priority_;
```
Could just add `priority_` to the tuple and negate it.
```
return std::tie(-lhs.priority_, lhs.library_,...) < std::tie(-rhs.priority_, rhs.library_,...)
```

mellery451 replied:
I like the suggestion, but I think I need to switch to `make_tuple` in that case (?) b/c `tie` expects lvalues.

seelabs replied:
I don't think I'd use `make_tuple` because it's going to be too easy to make a copy of member variables. We could assign the negative value to a local, but I'm not sure that's any better than what's there now. I'd probably keep what's there.

seelabs replied:
Actually, I think `std::forward_as_tuple` would work.

seelabs commented on src/ripple/app/main/Main.cpp:
```diff
@@ -181,6 +181,39 @@ void printHelp (const po::options_description& desc)
+        selectors_.reserve(v.size());
+        std::for_each(v.begin(), v.end(),
+            [this](std::string const& s)
+            {
```
I expected to see code that would strip whitespace off the ends - either here or in the selector constructor. Do we need to do that? I'd probably strip it off here we need to.

mellery451 replied:
I agree we should trim here - will add

seelabs commented on src/ripple/app/main/Main.cpp:
```diff
@@ -181,6 +181,39 @@ void printHelp (const po::options_description& desc)
+class multi_selector
+{
+private:
+    std::vector <beast::unit_test::selector> selectors_;
```
Nit: Formatting, `std::vector<` (i.e. no space between type and template list).

scottschurr commented:
Just poking around a little bit.  It looks like the syntax for invoking unit tests is expanding a bit.  At the moment (not your fault) the in-tree documentation for how to _run_ the unit tests seems pretty weak.  All I could find was what `--help` returns:
```
  -u [ --unittest ] arg Perform unit tests.
  --unittest-arg arg    Supplies argument to unit tests.
  --unittest-log arg    Force unit test log output, even in quiet mode.
  --unittest-jobs arg   Number of unittest jobs to run.
  --unittest-child      For internal use only. Run the process as a unit test 
                        child process.
```
This is sufficient to let me run `--unittest`, but I really couldn't figure out any of the other choices from this.  What are valid `arg`?  How do I find out what those are?

Does it make sense to add, say, a smallish `*.md` file, maybe at the top of the `test` tree?  It could provide some longer examples of how to run the unit tests.  I think _I_ might find myself referring to such a document from time to time.  And, since we're expanding the syntax for running the tests, now might be a good time to add it.

Just a thought.

mellery451 commented:
@scottschurr I agree...that could be clearer. I'll try to succinctly improve the usage descriptions and I'll look into adding some sort of test README.

mellery451 commented:
@scottschurr I updated the help to look like this:

```
rippled [options] <command> <params>

General Options:
  --conf arg             Specify the configuration file.
  --debug                Enable normally suppressed debug logging
  --fg                   Run in the foreground.
  -h [ --help ]          Display this message.
  --quorum arg           Override the minimum validation quorum.
  --silent               No output to the console after startup.
  -a [ --standalone ]    Run with no peers.
  -v [ --verbose ]       Verbose logging.
  --version              Display the build version.

RPC Client Options:
  --rpc                  Perform rpc command - see below for available
                         commands. This is assumed if any positional parameters
                         are provided.
  --rpc_ip arg           Specify the IP address for RPC command. Format:
                         <ip-address>[':'<port-number>]
  --rpc_port arg         Specify the port number for RPC command.

Ledger/Data Options:
  --import               Import an existing node database (specified in the
                         [import_db] configuration file section) into the
                         current node database (specified in the [node_db]
                         configuration file section).
  --ledger arg           Load the specified ledger and start from the value
                         given.
  --ledgerfile arg       Load the specified ledger file.
  --load                 Load the current ledger from the local DB.
  --net                  Get the initial ledger from the network.
  --nodetoshard          Import node store into shards
  --replay               Replay a ledger close.
  --start                Start from a fresh Ledger.
  --valid                Consider the initial ledger a valid network ledger.
  --validateShards       Validate an existing shard database (specified in the
                         [shard_db] configuration file section).

Unit Test Options:
  -q [ --quiet ]         Suppress test suite messages, including suite/case
                         name (at start) and test log messages.
  -u [ --unittest ] arg  Perform unit tests. The optional argument specifies
                         one or more comma-separated selectors. Each selector
                         specifies a suite name, full-name (lib.module.suite),
                         module, or library (checked in that order).
  --unittest-arg arg     Supplies an argument string to unit tests. If
                         provided, this argument is made available to each
                         suite that runs. Interpretation of the argument is
                         handled individually by any suite that accesses it --
                         as such, it typically only make sense to provide this
                         when running a single suite.
  --unittest-log         Force unit test log message output. Only useful in
                         combination with --quiet, in which case log messages
                         will print but suite/case names will not.
  --unittest-jobs arg    Number of unittest jobs to run in parallel (child
                         processes).
```

let me know what you think. I also added a short README file which may or may not be helpful.

scottschurr commented on src/test/README.md:
```diff
@@ -0,0 +1,30 @@
+
+# Unit Tests
```
Speaking only for myself, I think this file is really useful.  Thanks.

scottschurr commented on src/test/README.md:
```diff
@@ -0,0 +1,30 @@
+just `NoRippleCheckLimits`).
+
+More than one suite or group of suites can be specified as a comma separated
+list via the argument. For example, `--unittest=beast,OversizeMeta` will run
```
I'm noting that the examples in this file use `=` with no spaces.  But the `--help` says to use spaces in the arguments.  I tried both.  They both work.  It might be good, however, for our messaging to be consistent.

My personal preference is to use an `=` as the first argument separator and commas (with no spaces) as the remaining separators.  Just like you've shown here.  That visually glues all of the options together.  I haven't really paid any attention to boost::program_options.  I bet it would be hard to get it to show `=` instead of spaces.  Bummer.  It's not worth a lot of messing around.

scottschurr commented:
Oh.  And I think the reorganizing you did with the `--help` output looks good.

seelabs commented:
In 1.1.0-b2
